staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go

```diff
@@ -0,0 +1,249 @@
+					ObjectMeta: metav1.ObjectMeta{
+						Name: plName,
+					},
+					Spec: flowcontrol.PriorityLevelConfigurationSpec{
+						Type: flowcontrol.PriorityLevelEnablementLimited,
+						Limited: &flowcontrol.LimitedPriorityLevelConfiguration{
+							NominalConcurrencyShares: 100,
+							LendablePercent:          &testCase.constraints[flow].lendable,
+							BorrowingLimitPercent:    &testCase.constraints[flow].borrowing,
+							LimitResponse: flowcontrol.LimitResponse{
+								Type: flowcontrol.LimitResponseTypeQueue,
+								Queuing: &flowcontrol.QueuingConfiguration{
+									Queues:           10,
+									HandSize:         2,
+									QueueLengthLimit: 10,
+								},
+							},
+						},
+					},
+				}
+				cfgObjs = append(cfgObjs, fsObjs[flow], plcObjs[flow])
+			}
+			clientset := clientsetfake.NewSimpleClientset(cfgObjs...)
+			informerFactory := informers.NewSharedInformerFactory(clientset, time.Second)
+			flowcontrolClient := clientset.FlowcontrolV1beta3()
+			clk := eventclock.Real{}
+			controller := newTestableController(TestableConfig{
+				Name:                   "Controller",
+				Clock:                  clk,
+				AsFieldManager:         ConfigConsumerAsFieldManager,
+				FoundToDangling:        func(found bool) bool { return !found },
+				InformerFactory:        informerFactory,
+				FlowcontrolClient:      flowcontrolClient,
+				ServerConcurrencyLimit: 24,
+				RequestWaitLimit:       time.Minute,
+				ReqsGaugeVec:           metrics.PriorityLevelConcurrencyGaugeVec,
+				ExecSeatsGaugeVec:      metrics.PriorityLevelExecutionSeatsGaugeVec,
+				QueueSetFactory:        fqs.NewQueueSetFactory(clk),
+			})
+
+			ctx, cancel := context.WithTimeout(context.Background(), 50*time.Second)
+			stopCh := ctx.Done()
+			controllerCompletionCh := make(chan error)
+
+			informerFactory.Start(stopCh)
+
+			status := informerFactory.WaitForCacheSync(ctx.Done())
+			if names := unsynced(status); len(names) > 0 {
+				t.Fatalf("WaitForCacheSync did not successfully complete, resources=%#v", names)
+			}
+
+			go func() {
+				controllerCompletionCh <- controller.Run(stopCh)
+			}()
+
+			// ensure that the controller has run its first loop.
+			err := wait.PollImmediate(100*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+				return controller.hasPriorityLevelState(plcObjs[0].Name), nil
+			})
+			if err != nil {
+				t.Errorf("expected the controller to reconcile the priority level configuration object: %s, error: %s", plcObjs[0].Name, err)
+			}
+
+			concIntegrators := make([]fq.Integrator, 2)
+			reqInfo := &request.RequestInfo{
+				IsResourceRequest: false,
+				Path:              "/foobar",
+				Verb:              "GET",
+			}
+			noteFn := func(fs *flowcontrol.FlowSchema, plc *flowcontrol.PriorityLevelConfiguration, fd string) {}
+			workEstr := func() fcrequest.WorkEstimate { return fcrequest.WorkEstimate{InitialSeats: 1} }
+			qnf := fq.QueueNoteFn(func(bool) {})
+			var wg sync.WaitGroup
+			wg.Add(30)
```

Comment:
Did you mean to also have a `defer wg.Done()` here or something like that?

Reply:
I think I have read the whole thing now and this is the only thing I saw.

Reply:
The corresponding `wg.Done()` is inside the goroutine inside the loop below, at line 204.

Reply:
Sorry, I meant to type `Wait()`

Reply:
(Add and Done are called, but AFAICT not Wait, which makes it pointless)

Reply:
The force-push to 7c06996bb9d moves that `wg.Done()` to the top of the goroutine, with a `defer`.

Reply:
Still need to call `wg.Wait()` somewhere?

Reply:
D'Oh!

The previous force-push did not actually change the Done, I forgot to commit.
The force-push to a36f07e6ad6 changes the Done and adds the Wait.

Reply:
Now that I look at it again, this calls Add(30) but I had to do math to ensure that Done is actually called that many times. Ordinarily I'd ask to put Add(1) calls prior to starting the goroutines, but see my next comment because I'm not convinced that this waitgroup is actually needed, can you just remove it? Or modify to use it to make sure all the goroutines have exited before letting the test end (this is what I assumed it was for originally).